### DISCUSSION TITLE
Stereo Faust implementation

### DISCUSSION THREAD

**Discussion by eccessivo75** [pschatzmann/arduino-audio-tools#1003] [Q&A] [answered]
Hello,
I have implemented successfully a single channel  Faust effect (simple pitch shifter) on a Lyrat 4.3 board.
If  i try to use a stereo Faust Dsp, i would expect to have the signal processed separately on each channel, in order to preserve the channel separation. But on all my experiments it doesn't, i get the signal merged and processed in a dual mono style.
Am i doing something wrong? Or Faust is implemented as dual mono only ?
Thank you.

**Comment by pschatzmann** [comment, marked as answer]
I don't really know what you mean with dual mono. 
In Faust Stereo is represented as 2 dimensional array whereas the AudioTools expects interleaving left, right values.

You can review the [FaustStream](https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioLibs/AudioFaust.h) class which implements the conversion between these 2 concepts.

**Reply by eccessivo75** [reply]
Thank you for your answer.  So, left and right values are kept separated across all the path, but that doesn't happen to me . If i copy the stream from the input(kit) to the output(kit) left and right are completely separated , but immediately after i route the signal trough  the faust DSP left is merged with the right channel.

**Reply by pschatzmann** [reply]
Did you make a simple test with a stereo DSP that just copies stereo data ?

**Comment by pschatzmann** [comment]
Just added a test case and it works perfectly: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-faust/streams-faust_stereo_copy

I think you are doing something wrong: did you check  in your generated source that the input and output is 2 ?

**Reply by eccessivo75** [reply]
You are correct. Using your copy.h in my code works perfectly. So the problem was is in my faust code. 
I used the Faust web ide to modify the pitch shifter example for a version with 2 channels, but as soon as i use this in my code, i get only a loud noise on both the outputs. I am now checking now the faust compiler output file to understand what is wrong , but it is look simply a doubled version of the single channel version, nothing weird... has anyone experience the same behavior?

**Comment by pschatzmann** [comment]
I could think of the following:
- an integer overflow when Faust generates audio values > 1.0 in the output array. I suggest to add some asserts.
- floating point operations are quite slow on microcontroolers, so there is a risk that there are too many floating point operations so that we can not generate the result in time for the requested sample rate. Did you try to reduce the sample rate ?

**Reply by pschatzmann** [reply]
I committed a correction that clips the result from the Faust compute() result

**Reply by eccessivo75** [reply]
Yes, i tried decreasing the value of the sample rate . Even at 8000  i get the same kind of noise with apparently with no difference for quality and magnitudo. The noise comes even with no input signal.
To verify if it could be a problem related to the lack of resources , i move the 2 processes from a parallel layout to a cascaded layout (pitchshifter1:pitchshifter2), and it started working correctly.
I will investigate now your first suggestion.
I attach the  Compiled dsp not actually working: by Using this stereops.h  inside the example that you previously provided  you'll be able to reproduce the noise
[stereops.zip](https://github.com/pschatzmann/arduino-audio-tools/files/12746387/stereops.zip)

**Comment by pschatzmann** [comment]
Maybe a good test case would be to feed it a sine signal and to print the result values in the compute() of the pitch shift. If the output is not a sine then the issue is in the Faust Code

**Comment by pschatzmann** [comment]
There was indeed something wrong with the stereo logic: I was passing the number of samples to the count parameter of the compute method, but it seems that the number of frames is expected. In mono, these values are the same...

**Reply by eccessivo75** [reply]
That would explain the behavior.  If get you correctly, the problem could come before the Faust code, and it's about the way some variables are passed to Faust?

**Reply by pschatzmann** [reply]
Just have a look at my [latest correction](https://github.com/pschatzmann/arduino-audio-tools/commit/b15cbd9fb3fa93cf4b763421a7e05011ee62f780)...

**Reply by eccessivo75** [reply]
I implemented the correction, and the issue was solved.  Thank you for providing a solution i don't think i would have any chance to get through this alone.
